AAI/FAQ.md

```diff
@@ -238,7 +238,157 @@ client <- clearing : Client is given data
+"verified" using public/private keys.
+
+
+### Why are the Visa claims formatted as JWTs inside the Passport?
+
+A section on the importance to the NiH of retaining original authority in visas.
+
+
+### What are the ways that key management is done in practice?
+
+There are a variety of approaches that can be used for key
+management for Passports and Visas. We will first detail those that can be used for Passports
+and then discuss some extra wrinkles for Visas.
+
+For this discussion we assume there is a concrete JWT from issuer `https://issuer.example.org`
+(possibly a Broker *or* a Visa Issuer).
+
+My clearing house service 'trusts' the above issuer to help make data
+access decisions - and that trust is stated probably through some sort of explicit
+configuration. For our example lets imagine it has a hypothetical YAML configuration file
+
+```yaml
+trusted_brokers:
+  - https://issuer.example.org
+  - https://login.broker.com
+
+trusted_visa_issuers:
+  - https://dac.gov.world
+``` 
+ 
+The service now wants to verify a Passport or Visa
+JWT purporting to be from that issuer.
+
+---
+
+#### Use `jku` in the header of the JWT
+
+The JKU is the URL of a JSON file containing the issuers' public keys.
+
+```json
+{
+  "iss": "https://issuer.example.org",
+  "jku": "https://issuer.example.org/public-keys.json",
+  "kid": "key-october-1"
+}
+```
+
+For our concrete example we say that it is a JSON file residing
+at `https://issuer.example.org/public-keys.json`. We should note that
+there is no requirement that the file is hosted at the same
+location as the issuer (it could have been at `https://keys.example.org` for instance).
+
+**IMPORTANTLY**, for the secure use of this key management technique - the JKU 
+**MUST** also be whitelisted as part of the configuration of **OUR** service.
+Maybe we expand out of service configuration file to include the JKU.
+
+```yaml
+trusted_brokers:
+  - issuer: https://issuer.example.org
+    jku: https://issuer.example.org/public-keys.json
+
+  - issuer: https://login.broker.com
+    jku: https://keys.broker.com/set
+```
+
+It is **NEVER** valid to even attempt to access a JKU from a JWT header - unless the URL
+is already known to belong to the given issuer. (linkt to attack)
+
+To verify a JWT, the content of the JKU file is loaded and the `kid` is
+looked up in key set. The signature is validated using the public key found.
+
+Although this configuration requires explicit registration of JKUs, the content
+of the key sets can allow the best practice of key rotation.
+
+The content of the JKU file is designed to be cached aggressively but the set of keys
+can evolve/rotate slowly (think weeks/months - not hours).
+
+
+---
+
+#### Use the `kid` in the header of the JWT along with OIDC Discovery
+
+The OIDC discovery protocol allows the use of the JWT issuer - in conjunction with a call
+to a .well-known/openid-configuration to look up the location of the public key file.
```

Comment:
A link to [OpenID Provider Metadata specification](https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata) defining the `jwks_uri` metadata key may be useful.

Reply:
I would suggest we emphasize the utility of global uniqueness of key ids.

Reply:
@kwrodarmer Key ids do not need to be globally unique. What needs to be unique is the tuple (JWKS URL, key id), so in the three cases discussed in the text, the first two cases are (jku,kid) and (jwks_uri,kid). Only the third case with keys preshared out-of-band have empty JWKS URL, so the tuple is ('',kid). Even in this case the key ids need to be unique within the receiving system only, not globally. But having them globally unique covers the case, so it may be worth mentioning.